orchagent/cbf/cbfnhgorch.cpp

```diff
@@ -0,0 +1,819 @@
+        if (success)
+        {
+            it = consumer.m_toSync.erase(it);
+        }
+        else
+        {
+            ++it;
+        }
+    }
+}
+
+/*
+ * Purpose: Validate the CBF members.
+ *
+ * Params:  IN members - The members string to validate.
+ *
+ * Returns: Pair where:
+ *          - the first element is a bool, representing if the members are
+ *            valid or not
+ *          - the second element is a vector of members
+ */
+pair<bool, vector<string>> CbfNhgOrch::getMembers(const string &members)
+{
+    SWSS_LOG_ENTER();
+
+    auto members_vec = swss::tokenize(members, ',');
+    set<string> members_set(members_vec.begin(), members_vec.end());
+    bool success = true;
+
+    /*
+     * Verify that the members list is not empty
+     */
+    if (members_set.empty())
+    {
+        SWSS_LOG_ERROR("CBF next hop group members list is empty.");
+        success = false;
+    }
+    /*
+     * Verify that the members are unique.
+     */
+    else if (members_set.size() != members_vec.size())
+    {
+        SWSS_LOG_ERROR("CBF next hop group members are not unique.");
+        success = false;
+    }
+
+    return {success, members_vec};
+}
+
+/*
+ * Purpose: Constructor.
+ *
+ * Params:  IN index - The index of the CBF NHG.
+ *          IN members - The members of the CBF NHG.
+ *          IN selection_map - The selection map index of the CBF NHG.
+ *
+ * Returns: Nothing.
+ */
+CbfNhg::CbfNhg(const string &index,
+               const vector<string> &members,
+               const string &selection_map) :
+    NhgCommon(index),
+    m_selection_map(selection_map)
+{
+    SWSS_LOG_ENTER();
+
+    uint8_t idx = 0;
+    for (const auto &member : members)
+    {
+        m_members.emplace(member, CbfNhgMember(member, idx++));
+    }
+}
+
+/*
+ * Purpose: Move constructor.
+ *
+ * Params:  IN cbf_nhg - The temporary object to construct from.
+ *
+ * Returns: Nothing.
+ */
+CbfNhg::CbfNhg(CbfNhg &&cbf_nhg) :
+    NhgCommon(move(cbf_nhg)),
+    m_selection_map(move(cbf_nhg.m_selection_map)),
+    m_temp_nhgs(move(cbf_nhg.m_temp_nhgs))
+{
+    SWSS_LOG_ENTER();
+}
+
+/*
+ * Purpose: Sync the CBF NHG over SAI, getting a SAI ID.
+ *
+ * Params:  None.
+ *
+ * Returns: true, if the operation was successful,
+ *          false, otherwise.
+ */
+bool CbfNhg::sync()
+{
+    SWSS_LOG_ENTER();
+
+    /* If the group is already synced, exit. */
+    if (isSynced())
+    {
+        return true;
+    }
+
+    /* Create the CBF next hop group over SAI. */
+    sai_attribute_t nhg_attr;
+    vector<sai_attribute_t> nhg_attrs;
+
+    nhg_attr.id = SAI_NEXT_HOP_GROUP_ATTR_TYPE;
+    nhg_attr.value.u32 = SAI_NEXT_HOP_GROUP_TYPE_CLASS_BASED;
+    nhg_attrs.push_back(move(nhg_attr));
+
+    /* Add the number of members. */
+    nhg_attr.id = SAI_NEXT_HOP_GROUP_ATTR_CONFIGURED_SIZE;
+    assert(m_members.size() <= UINT32_MAX);
+    nhg_attr.value.u32 = static_cast<sai_uint32_t>(m_members.size());
+    nhg_attrs.push_back(move(nhg_attr));
```

Comment:
Any restriction of number of members based on SAI_SWITCH_ATTR_MAX_NUMBER_OF_FORWARDING_CLASSES??

Reply:
The maximum number of forwarding classes doesn't affect how many members a group can have. It would admittedly be strange to have more group members than forwarding classes, but is allowable in the data plane and it could be that the user wants the ability to change maps while keeping the same members. I'll add a warning log to indicate that this may be a mistake by the user.

Reply:
Log added